src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Corax;
+using Corax.Queries;
+using Raven.Server.Documents.Indexes.Static.Spatial;
+using Raven.Server.Documents.Queries;
+using Raven.Server.Documents.Queries.Results;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Logging;
+using Voron.Impl;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public class CoraxIndexReadOperation : IndexReadOperationBase
+    {
+        private readonly IndexSearcher _indexSearcher;
+        private readonly CoraxQueryEvaluator _coraxQueryEvaluator;
+        private long _entriesCount = 0;
+        private const int _bufferSize = 2048;
+
+        public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTransaction) : base(index, logger)
+        {
+            _indexSearcher = new IndexSearcher(readTransaction);
+            _coraxQueryEvaluator = new CoraxQueryEvaluator(_indexSearcher);
+        }
+
+        public override void Dispose()
+        {
+            _indexSearcher?.Dispose();
+        }
+
+        public override long EntriesCount() => _entriesCount;
+
+        public override IEnumerable<QueryResult> Query(IndexQueryServerSide query, QueryTimingsScope queryTimings, FieldsToFetch fieldsToFetch,
+            Reference<int> totalResults, Reference<int> skippedResults,
+            IQueryResultRetriever retriever, DocumentsOperationContext documentsContext, Func<string, SpatialField> getSpatialField, CancellationToken token)
+        {
+            var pageSize = query.PageSize;
```

Comment:
This is `int32` - I believe that it is safe to assume that a _query_ limit of 2.1 billion entries is fine (separated from the index limit, which is much higher).
However, to be on the safe side, we probably want to make that a `long` anyway, since that would simplify things for docs / users

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-17215